include/swift/Syntax/Syntax.h

```diff
@@ -47,19 +63,137 @@ SyntaxNode makeRoot(const RawSyntax *Raw) {
 
 const auto NoParent = llvm::None;
 
-/// The main handle for syntax nodes - subclasses contain all public
-/// structured editing APIs.
+/// Marker type to construct \c SyntaxRef nodes without validation. This is used
+/// to create \c SyntaxRef inside \c OwnedSyntaxRef and \c
+/// OptionalOwnedSyntaxRef that point to a \c SyntaxDataRef which is yet to be
+/// initialised.
+/// Validation will occur in these types once the \c SyntaxRef is accessed.
+struct no_validation_t {};
+
+// MARK: - OwnedSyntaxRef
+
+/// Holds a \c SyntaxDataRef and provides a \c SyntaxRef (or one of its
+/// subclasses) as an accessor to the \c SyntaxDataRef.
+/// The user of this type needs to make sure that the \c OwnedSyntaxRef always
+/// outlives the \c SyntaxRef provided by it, because otherwise the \c SyntaxRef
+/// points to invalid memory.
+/// It allows transparent access to the \c SyntaxRef through the \c -> operator.
+///
+/// All methods that return a \c OwnedSyntaxRef should be inlined to avoid
+/// copying the \c SyntaxDataRef, which is rather expensive because the struct
+/// is rather large.
 ///
-/// Essentially, this is a wrapper around \c SyntaxData that provides
-/// convenience methods based on the node's kind.
+/// A typical initialisation of a OwnedSyntaxRef looks as follows:
+/// \code
+/// OwnedSyntaxRef<MySyntaxRef> Result;
+/// someSyntaxDataRef.getChildRef(Index, Result.getDataPtr());
+/// \endcode
+/// The first line creates an empty \c OwnedSyntaxRef with uninitialised memory.
+/// The second line invokes a method that fills \c Data of \c OwnedSyntaxRef.
+/// This way, we directly write the \c SyntaxDataRef to the correct memory
+/// location and avoid copying it around.
+template <typename SyntaxRefType>
+class OwnedSyntaxRef {
+  SyntaxDataRef Data;
+  SyntaxRefType Ref;
+
+public:
+  /// Create an *uninintialized* \c OwnedSyntaxRef. Its storage needs to be
+  /// initialised by writing a \c SyntaxDataRef to the pointer returned by
+  /// \c getDataPtr()
+  /// Implementation Note: We need to initialise \c Ref without validation,
+  /// because \c Data is still uninitialised. \c Ref will be validated when
+  /// accessed using \c getRef or \c -> .
+  OwnedSyntaxRef() : Data(), Ref(getDataPtr(), no_validation_t()) {}
+
+  OwnedSyntaxRef(const OwnedSyntaxRef &Other)
+      : Data(Other.Data), Ref(getDataPtr(), no_validation_t()) {}
+  OwnedSyntaxRef(OwnedSyntaxRef &&Other)
+      : Data(std::move(Other.Data)), Ref(getDataPtr(), no_validation_t()) {}
+
+  /// The pointer to the location at which \c this stores the \c Data.
+  /// Can be used to retroactively populate the \c Data after \c OwnedSyntaxRef
+  /// has been constructed with uninitialised memory.
+  SyntaxDataRef *getDataPtr() { return &Data; }
+
+  const SyntaxRefType &getRef() {
+    assert(Ref.getDataRef() == getDataPtr() &&
+           "Ref no longer pointing to Data?");
```

Comment:
If `Data` is const, this `assert` is not needed.

Reply:
I ran into issues here at some point after moving `OwnedSyntaxRef` to a different location. The pointer stays constant but the `SyntaxDataRef` now lives in a different location, so the pointer isn’t valid anymore. That’s why I put the assertion in and I think it’s still useful if someone starts to fiddle around `OwnedSyntaxRef`.